### DISCUSSION TITLE
Decoding scoreboard RF data

### DISCUSSION THREAD

**Discussion by ax25** [merbanan/rtl_433#2392] [Ideas]
Hello!

I'd like to ask the community if this would be possible to do using the rtl_433 framework:

I successfully built a simple scoreboard decoder to read data from several scoreboard controllers for sports from different brands. I used gnuradio to build the demodulator and the decoder. The signals are typically 2FSK located around 433MHz or 868MHz, with baudrates of 4800 or 9600. The data is sent continuously or in bursts using packets of different lengths and payload depending on the data that is being transmitted.

As the protocols are not published, normally I take samples of the signals, use inspectrum to find patterns (like preambles), and build the decoder in gnuradio to capture raw data. Then I look for changes between what should be data (changes in the time, score, etcetera), and I build a decoder to get time, scores, faults... I did not program any of the checksums and the result is not robust.

This way I built a decoder for some brands that are common in my country (Spain). This is being used for live streaming purposes, basically connecting the device (currently a raspberry pi and an rtlsdr) via socket to a server that has the ability to overlay graphics to the video feed.

A scoreboard controller is not a sensor by any means. However, could rtl_433 be adapted easily to decode these types of signals?

What I would like to achieve by moving to rtl_433 would be:

1) To have a solid demodulator and decoder structure.
2) To open the project to the community, so the scoreboard decoder can be used for live streaming purposes widely.
3) To be able to share recordings of signals with the community with the intention to add compatibility with more brands.

What I can provide:

- Recordings of some signals (using gnuradio)
- Some reverse-engineered decoders in python (with my limited coding skills)

My knowledge of C is limited and I am not able to find where to start looking in the rtl_433 sources. For instance, I am able to find where the different decoders are written, but I can't find the demodulators. For me, it always starts with a preamble! But I can't find any in the sources.

If anyone finds this could be interesting or doable, please let me know.

**Comment by zuckschwerdt** [comment]
Yes, this will work. These should be the same kind of transmissions rtl_433 usually processes.

Upload some samples (as zip) here and we'll take a look.

You only need the decoder, everything from framing, demod to bit slicers is automatic. If you know the bitrate you can start with a flex decoder: try `-A` on a sample and use the `-X` suggestion.

**Reply by ax25** [reply]
I have a scoreboard controller with me to test. It is generating a signal at 433.517MHz with a deviation frequency of +/-45kHz. For this model, the data is being transmitted in bursts (there is RF silence between messages) and the data rate is 4800baud. 

I'm trying to decode using the command:

% rtl_433 -f 433517000 -A -s 1000000 

After a few seconds it starts to spit some decoded messages. The -X suggestion is:

-X 'n=name,m=FSK_PCM,s=208,l=208,r=212992'

The decoded data is showing some constant data followed by what seams random data. The s=208 and l=208 make sense (bit duration 208us is consistent with 4800baud). I don't know what the r parameter means.

What I can add is that I know that there is some sort of preamble = [1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0] that is working fine for me in my python decoder.

The decoded message shows messages like this one:

`time      : 2023-02-22 22:54:23
model     : name         count     : 1             num_rows  : 1             rows      : 
len       : 183          data      : aaaaaaaaa96ea30b6ba3284992020aa1818981819bea40
codes     : {183}aaaaaaaaa96ea30b6ba3284992020aa1818981819bea40`

Which is consistent with the preamble I just mentioned.

May I ask what command should I use to generate the recordings?

**Comment by merbanan** [comment]
-S unknown.

**Comment by ax25** [comment]
I was able to record samples at 500ksps.
[433.517M_500k.zip](https://github.com/merbanan/rtl_433/files/10809154/433.517M_500k.zip)

Time is running (around 02:00 or so and counting) and score is 01-02.

**Reply by zuckschwerdt** [reply]
Note that `500k` did not work: `Invalid sample rate: 500000 Hz [sdr_set_sample_rate] Failed to set sample rate.` You now have some other random sample rate, 2048k I guess.

**Comment by merbanan** [comment]
src/rtl_433 -s 500k -r ../signals/scoreboard/g004_433.517M_500k.cu8 -X 'n=name,m=FSK_PCM,s=840,l=840,r=21200'
```

{183}2aaaaaaaa96ea30b6ba3284992020aa18189819198d950
{183}eaaaaaaaa96ea30b6ba3284992020a61b9b989f9986e68
```

The sample signals are a little clipped.

**Comment by merbanan** [comment]
And I dont really see any pattern.

**Comment by ax25** [comment]
Same output that I got, with changing values at the end of the data.

In gnuradio I am able to read data with the devices next to each other and default RTL gain.

**Comment by ax25** [comment]
The decoder might be taking a couple of extra bits prior to the preamble.

2aaaaaaaa (0010 1010 1010....) should be aaaaaaaa (1010 1010...).

The way I coded it was:

- look for the preamble [1,0,1,0, 1,0,1,0, 1,0,1,0, 1,0,1,0, 1,0,1,0, 1,0,1,0, 1,0,1,0, 1,0,1,0, 1,0,1,0, 0,1,0,1] or aaaaaaaaa5
- when found, start decoding. So, if there are extra bits before the preamble it wouldn't change the outcome.

Can we force the preamble in the rtl_433 call?

**Reply by zuckschwerdt** [reply]
Yes, add `,preamble=aaa5`

**Comment by ax25** [comment]
I'm definitely getting closer.

There is a part of the data that is never changing, like a very large preamble. After that, the next digit might be an indication of which kind of value is being delivered.

So:

rtl_433 -f 433517000 -s 1M -R 0 -X 'n=name,m=FSK_PCM,s=208,l=208,r=212992,preamble={120}aaaaaaaaa5ba8c2dae8ca126480828'

The result is still not readable, but it is changing every second (and stops if I stop the time), which is an indication that there is time data being transmitted.

This is exciting.

**Comment by ax25** [comment]
I got the time.

-s 1M -R 0 -X 'n=name,m=FSK_PCM,s=208,l=208,r=212992,preamble={20}04143'

Result:

{58}313730333377160

there is always a "3" as padding between data. In this example time is 17 minutes 03 seconds. Last part is checksum, and I'm afraid it will be hard to reverse engineer that part.

I was able to decode the time in a couple of hours, while it took me literally a dozen of hours to find out manually.

This looks so promising!

Edit: the problem in the first data with the "aaaa" preamble is that there is a couple of extra bits that are shifted and it affects the whole decoded message. Choosing the right preamble seems to fix it.

Edit 2: the right preamble seems to be some sort of 555 instead of aaa

Edit 3: I was able to get the score too.
-f 433517000 -s 1M -R 0 -X 'n=name,m=FSK_PCM,s=208,l=208,r=212992,preamble={16}0415'

**Comment by merbanan** [comment]
https://triq.net/bitbench#c=%7B183%7D2aaaaaaaa96ea30b6ba3284992020aa18189819198d950&c=%7B183%7Deaaaaaaaa96ea30b6ba3284992020a61b9b989f9986e68&c=%7B183%7Daaaaaaaaa96ea30b6ba3284992020aa1818981819bea40&f=%3F%3F%3Ahh%20MH%3Ah%20P%3Ah%20ML%3Ah%20P%3Ah%20SH%3Ah%20P%3Ah%20SL%3Ah%20unk%3Ahhhhhhh&a=Preamble&m=0415&cw=4

**Comment by merbanan** [comment]
You can use bit bench to figure out the meaning of bits. Can you pin point where you found the score ?

**Comment by ax25** [comment]
MH, ML local
SH, SL away

01 - 02

**Comment by ax25** [comment]
The preamble I chose is just a way to trim the data and be able to read values, but it is not a preamble, which can be problematic. For instance, I realized there is a button in the controller to turn on the "horn" and it affects the 0->1 in this chosen preamble. 

I find hard to understand why there is so much static data generated by this controller... I guess there has to be some serial number and sport type or something like that.

This specific controller is one of the simplest examples. It only works for football, so data is small and being transmitted at a low rate.

Tried to run rtl_433 with all the decoders and this one and it is great to see how it is able to pick temperature sensors and the score board at the same time. I chose 1Msps.

**Comment by ax25** [comment]
From [](https://triq.org/rtl_433/PULSE_FORMATS.html)

`an optional "sync word" to identify the start of payload data, a very common sync word is 2dd4`

There is in fact this sync word in the message.

**Reply by zuckschwerdt** [reply]
Good catch. Use a preamble of `55552dd4` (full preable is `5555555552dd4`). Not you got alignment and polarity of the payload data correct.

**Reply by ax25** [reply]
what about checksum (or similar) calculation? Do you think it is possible to find out how the last data digits are being calculated in order to have data consistency?

Some examples:

{130}616d74650932404143343632303319c60
{130}616d746509324041433436313933fa0e0
{130}616d746509324041433436313833c93f0

With data highlighted:
{130}616d746509324041433 `4` 3 `6` 3 `2` 3 `0` 33` 19c6` 0 -> time 46:20
{130}616d746509324041433 `4` 3 `6` 3 `1` 3 `9` 33 `fa0e` 0 -> time 46:19
{130}616d746509324041433 `4` 3 `6` 3 `1` 3 `8` 33 `c93f` 0 -> time 46:18

**Reply by zuckschwerdt** [reply]
Yes, that's CRC-16, poly 0x1021, init=0.

**Reply by merbanan** [reply]
Can you highlight the score also ? If we get that we can write a decoder.

**Reply by ax25** [reply]
{130}616d74650932404154303230323380f60
{130}616d746509324041543032303133d5a50
{130}616d746509324041543032303033e6940

Highlighted:
{130}616d746509324041543`0`3`2`3`0`3`2`33`80f6`0 -> 2-2
{130}616d746509324041543`0`3`2`3`0`3`1`33`d5a5`0 -> 2-1
{130}616d746509324041543`0`3`2`3`0`3`0`33`e694`0 -> 2-0

**Reply by ax25** [reply]
> Yes, that's CRC-16, poly 0x1021, init=0.

I am trying to calculate it using different parts of the payload but I can't get the right result.

**Reply by zuckschwerdt** [reply]
Use the whole payload. I.e. pseudo-code `CRC-16(616d74650932404154303230323380f6) == 0`

**Comment by ax25** [comment]
I understood why there is so much invariable data in the messages. It is coded in ASCII.

`amte	2@AC46203`
`amte	2@AC46193`
`amte	2@AC46183`

AC means Time. If AT is sent, it means Score.

This protocol was built by the scoreboard engineer, so it is highly non-standard stuff.

Every brand, every model, every country will have different signal modulation, encoding, and it is highly possible that even the same manufacturer and model have protocol variations.

By now I was able to decode a bunch of brands/models that are working near my city. For each new sport I try to build the decoder, so now I have been able to work with 10 different models or so.

**Reply by zuckschwerdt** [reply]
This is officially the first instance where [BitBenchs ASCII output](https://triq.net/bitbench#c=616d74650932404154303230323380f6&c=616d746509324041543032303133d5a5&c=616d746509324041543032303033e694&f=cccc%208h%20cccc%20ccccc%20CRC%3A16h) has ever been useful ;)

**Reply by ax25** [reply]
The CRC that is shown is just a representation or is it calculated?

**Reply by merbanan** [reply]
https://crccalc.com/?crc=616d746509324041543032303233&method=crc16&datatype=hex&outtype=0

The XMODEM version is the correct CRC-16 of 616d746509324041543032303233 [80f6]

**Reply by zuckschwerdt** [reply]
> The CRC that is shown is just a representation or is it calculated?

In BitBench? There is no CRC check or calculation in BitBench right now. It's planned ;)

And given the preamble, sync, and CRC I'd say the controller uses a CC1101 or similar. That explains the "dumb" ASCII encoding. Radio bridges like the CC1101 make it too easy, "designers" without radio knowledge use them (poorly).

**Reply by ax25** [reply]
> > The CRC that is shown is just a representation or is it calculated?
> 
> In BitBench? There is no CRC check or calculation in BitBench right now. It's planned ;)
> 
> And given the preamble, sync, and CRC I'd say the controller uses a CC1101 or similar. That explains the "dumb" ASCII encoding. Radio bridges like the CC1101 make it too easy, "designers" without radio knowledge use them (poorly).

I might be able to open it and look for the radio transmitter (or transceiver).

**Reply by zuckschwerdt** [reply]
If you a curious ;) Not really needed, but would be interesting to confirm.

**Reply by ax25** [reply]
@zuckschwerdt 

It is this one:

https://i.postimg.cc/RhGwhLrp/Whats-App-Image-2023-02-23-at-14-14-57.jpg

https://www.sparkfun.com/datasheets/Wireless/General/RFM22.PDF

**Comment by ax25** [comment]
FYI, this is this particular scoreboard controller:

https://i.postimg.cc/hh4MZYrr/Whats-App-Image-2023-02-23-at-13-27-33.jpg

**Reply by zuckschwerdt** [reply]
That's a HopeRF RF22 as RFM22 module. It's even easier than a CC1101.

@merbanan we had a request about RFM22 support (simple UART bridge) recently IIRC?

**Reply by merbanan** [reply]
That was a HC-12 device IIRC.

**Reply by merbanan** [reply]
Or this issue [Somfy io-homecontrol](https://github.com/merbanan/rtl_433/issues/1376)

**Reply by zuckschwerdt** [reply]
Ty.

Oh, and looking at the RFM22 manual, we expect 0-4 bytes header and 0-1 byte length. This matches the packet, e.g.
`HDR:cccc LEN:8h DATA:ccccccccc CRC:16h` see [this BitBench](https://triq.net/bitbench#c=616d74650932404154303230323380f6&c=616d746509324041543032303133d5a5&c=616d746509324041543032303033e694&f=HDR%3Acccc%20LEN%3A8h%20DATA%3Accccccccc%20CRC%3A16h).

So there is the fixed prefix explained :)

**Comment by ax25** [comment]
To add more data:

[This BitBench](https://triq.net/bitbench#c=616d7465093241414c3737313f334a1e&c=616d7465093241415430303030334c2f&c=616d746509324141433030313333fc89&f=HDR%3Acccc%20LEN%3A8h%20DATA%3Accccccccc%20CRC%3A16h&cw=4) shows AT (score), AC (time) and AL (brightness). Also, horn is always on, so instead of 2@ we can see 2A.

The AL (brightness) message includes:
Bytes 5, 6: Brightness value for time, score.
Byte 7: Period in play (1st, 2nd, 3rd...)
Byte 8: Hundred points (?=empty, 1=local, 2=away, 3=both).


If I have a decoder example to use as a starting point, I will be able to code some application level program in order to, for example, generate json data to feed the streaming software (vMix, OBS, or even ffmpeg).

**Reply by zuckschwerdt** [reply]
The link has old data. You need to enter data, then click "Link to this data and format" to get the fresh data as link.

**Reply by ax25** [reply]
you are right, now it is fixed

**Comment by ax25** [comment]
One question: will it work with a continuous signal? This particular controller has TDD transceiver (hence the rf silences between samples), but most of them use constant transmission. Is the demodulator/decoder ready for this kind of signals?

**Reply by merbanan** [reply]
First that is misuse of the SRD bands. Then we have rtl_433 that will segment the signal to a max length and then process it. So it will most likely work but some messages will be missed. But if there are constant repeats of the information then it will probably work in practice.

**Reply by ax25** [reply]
OK, we will try and see. I see some potential problems:
- for clock data, if there are large buffers the delay might be large and noticeable.
- tenths of seconds will get lost if there are lost packets.

I have some recordings using gnuradio file sinks. Is there a way to read them in rtl_433? 
Edit: I just tried it and it rtl_433 seems to open it without issues.
Edit2: well, I also tried with a .mp4 movie and it also seems to open it! haha.. will keep trying

**Comment by ax25** [comment]
I've been able to record a different scoreboard controller, from a different brand.

[Scoreboard-ContTX-4800bps.zip](https://github.com/merbanan/rtl_433/files/10833602/Scoreboard-ContTX-4800bps.zip)

433.075MHz FSK 4800baud, continuous TX

(EDIT: Frequency is 434.075MHz)

The -X guess is: `-X 'n=name,m=FSK_PCM,s=208,l=208,r=212992'`

Which is consistent with the 4800baud.

However, it seems like the demodulator waits for the whole buffer (very big) to be full before trying to decode bits.

Was able to decode [1 big data packet](https://triq.org/pdv/#AAB0130801019C00D0053C026C03BC08580BBC00008455+AAB0140801019C00D0053C026C03BC08580BBC0000919455+AAB0140801019C00D0053C026C03BC08580BBC000090A255+AAB0140801019C00D0053C026C03BC08580BBC000080B655+AAB0190801019C00D0053C026C03BC08580BBC0000808193A190C38455+AAB0180801019C00D0053C026C03BC08580BBC000080908080908455+AAB0170801019C00D0053C026C03BC08580BBC0000B19080C08455+AAB0130801019C00D0053C026C03BC08580BBC00009255+AAB0150801019C00D0053C026C03BC08580BBC000080919455+AAB0130801019C00D0053C026C03BC08580BBC00009455+AAB0130801019C00D0053C026C03BC08580BBC0000A455+AAB0140801019C00D0053C026C03BC08580BBC000080B655+AAB0130806019C00D0053C026C03BC08580BBC00009555+AAB0210801019C00D0053C026C03BC08580BBC0000D3A190C083C080C080908080C0809255+AAB0170801019C00D0053C026C03BC08580BBC000083818390A455+AAB0130801019C00D0053C026C03BC08580BBC0000A655+AAB0220801019C00D0053C026C03BC08580BBC0000C183A190C080C083C080908080C0809255+AAB0170801019C00D0053C026C03BC08580BBC000080818393A455+AAB0130801019C00D0053C026C03BC08580BBC0000A655+AAB01C0801019C00D0053C026C03BC08580BBC0000A193A190C080C080C08455+AAB0170801019C00D0053C026C03BC08580BBC0000908080C08455+AAB0130801019C00D0053C026C03BC08580BBC00009255+AAB0160801019C00D0053C026C03BC08580BBC00008081839455+AAB0130801019C00D0053C026C03BC08580BBC0000A455+AAB0130801019C00D0053C026C03BC08580BBC0000A655+AAB0130807019C00D0053C026C03BC08580BBC00009555+AAB0150801019C00D0053C026C03BC08580BBC00009091B455+AAB0150801019C00D0053C026C03BC08580BBC0000A1909255+AAB0130801019C00D0053C026C03BC08580BBC00008455+AAB0130801019C00D0053C026C03BC08580BBC00009255+AAB0140801019C00D0053C026C03BC08580BBC0000809255+AAB0130801019C00D0053C026C03BC08580BBC00008455).

Right now I have no info about possible preambles or data, but what worries me is the ability or not to decode continuous transmissions...

**Reply by zuckschwerdt** [reply]
Regarding legality: This is not 433.92 MHz (ISM / SRD) but LPD433 channel 41, which is legal in Spain for voice applications.

**Reply by ax25** [reply]
Oh, edited the frequency, it is actually 434.075MHz.

These frequencies are abundant in Spain when looking for RF scoreboard signals.

**Reply by ax25** [reply]
I'm sitting next to this controller (in a pavilion during a game, I can't take it home) and the rtl_433 demodulator can't seem to find any signal.

It is a FSK centered on 434.075, 4800baud, deviation frequency of about +/-6.5kHz.

<img width="1276" alt="Captura de pantalla 2023-02-26 a las 13 36 02" src="https://user-images.githubusercontent.com/22448360/221410854-4efdba66-73c2-4587-bdd0-802e45415280.png">

The preamble has to be some sort of groups of 10000000001000000000... or very similar if IIRC.

I'm trying rtl_433 -f 434075000 -A -R 0

Should I try anything different?

**Reply by zuckschwerdt** [reply]
rtl_433 can work well with continuous signals. We might support that some day, but currently it won't give good results.

**Reply by ax25** [reply]
I understand. It was a coincidence that the controller that I had with me was using time division multiplexing and gave good results right away. Almost 100% of the controllers use continuous tx.

**Comment by nonsensation** [comment]
@ax25 Hello, did you progress any further with this?

I am looking for exactly such a solution right now and was searching the internet all over - this seems to be the only mention of using an SDR stick to intercept common wireless scoreboards (433/868 MHz).

I am still at the very beginning, having bought only an RTL-SDR stick and did not yet learn enough to even get the details of this thread, but if there is a solution already that saves me hours of duplicated work I'd be very happy. If not, I'd like to work on this further.
Currently there is no solution to this, all teams that I follow on streams in my sport use a separate timer that they try to keep in sync manually..

**Comment by ax25** [comment]
I programmed my own solution and customized it for every scoreboard that I
work with. rtl_433 did not work because it is not ready for continuous
transmission.

**Reply by nonsensation** [reply]
Thank you for lettig me know.
I guess it is not available for public use?
Another question, in germany boards vfrom Stramatel are very common, did you work with this brand by coincidence?
